File: wma_crossover.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"strconv"
)

type KrakenResponse struct {
	Result map[string]interface{} `json:"result"`
}

func calculateWMA(prices []float64, period int) float64 {
	if len(prices) < period {
		return 0.0
	}
	subset := prices[len(prices)-period:]
	var sumWeights, weightedSum float64
	for i, price := range subset {
		weight := float64(i + 1)
		weightedSum += price * weight
		sumWeights += weight
	}
	return weightedSum / sumWeights
}

func wmaCrossoverSample() {
	symbol := "XXBTZUSD" // Kraken's pair name for BTC/USD
	url := fmt.Sprintf("https://api.kraken.com/0/public/OHLC?pair=%s&interval=1440", symbol)

	resp, _ := http.Get(url)
	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)

	var data KrakenResponse
	json.Unmarshal(body, &data)

	// Kraken returns data in nested arrays: [time, open, high, low, close, ...]
	ohlcData := data.Result[symbol].([]interface{})
	var closePrices []float64

	for _, bar := range ohlcData {
		b := bar.([]interface{})
		price, _ := strconv.ParseFloat(b[4].(string), 64)
		closePrices = append(closePrices, price)
	}

	fastWMA := calculateWMA(closePrices, 7)
	slowWMA := calculateWMA(closePrices, 30)
	currentPrice := closePrices[len(closePrices)-1]

	fmt.Printf("Current BTC Price: $%.2f\n", currentPrice)
	fmt.Printf("7-Day WMA: %.2f | 30-Day WMA: %.2f\n", fastWMA, slowWMA)

	if fastWMA > slowWMA {
		fmt.Println("Signal: 🚀 BULLISH (Golden Cross)")
	} else {
		fmt.Println("Signal: ⚠️ BEARISH (Death Cross)")
	}
}

File: wma_crossover.py
import ccxt
import pandas as pd

def get_crossover_signal(symbol='BTC/USD', fast_p=7, slow_p=30):
    exchange = ccxt.kraken({'enableRateLimit': True})
    
    # Fetch enough data to cover the slow period
    bars = exchange.fetch_ohlcv(symbol, timeframe='1d', limit=slow_p + 10)
    df = pd.DataFrame(bars, columns=['ts', 'open', 'high', 'low', 'close', 'volume'])

    def calculate_wma(prices, period):
        weights = list(range(1, period + 1))
        return (prices[-period:] * weights).sum() / sum(weights)

    # Calculate both WMAs
    fast_wma = calculate_wma(df['close'], fast_p)
    slow_wma = calculate_wma(df['close'], slow_p)
    current_price = df['close'].iloc[-1]

    # Signal Logic
    if fast_wma > slow_wma:
        signal = "🚀 BULLISH (Golden Cross)"
    elif fast_wma < slow_wma:
        signal = "⚠️ BEARISH (Death Cross)"
    else:
        signal = "NEUTRAL"

    return current_price, fast_wma, slow_wma, signal

if __name__ == "__main__":
    price, fast, slow, signal = get_crossover_signal()
    print(f"Price: ${price:,.2f} | Fast WMA: {fast:.2f} | Slow WMA: {slow:.2f}")
    print(f"Signal: {signal}")
